File: validation/check_localize.py
#!/usr/bin/env python3
"""Compare our Boys localization against PySCF.

    ./build/check_localize && python3 validation/check_localize.py

**Comparing localizations needs care, and the obvious comparison is wrong.** A
Boys solution is defined only up to the ordering and sign of the localized
orbitals, so coefficient matrices cannot be compared elementwise. Worse, the
functional has many local maxima and the PySCF optimizer is strongly start
dependent: on water/STO-3G its own cost function gives 8.641 for the canonical
orbitals, 6.271 from its default init_guess="atomic", and 7.699 from the canonical
orbitals with that guess disabled. So "does our number equal their number" is not
a test of correctness -- it is a test of whose starting point was luckier.

What is a test, and is the sharp one:

1. **Stationarity under the PySCF cost function.** Hand PySCF our orbitals with
   its initial guess disabled and let it optimize. If it cannot improve on them,
   they are a genuine stationary point of the objective PySCF itself defines,
   which is a stronger statement than agreeing with one of its local maxima.
2. **Quality.** Our spread should be no worse than what PySCF reaches on its own.
   On these cases it is better, which is the right way round -- but that is a
   property of the starting point rather than of correctness, so it is reported
   and not asserted.

The PySCF cost function is r2 - sum_i |<i|r|i>|^2, the sum of orbital spreads,
which it minimizes. That is the same objective this module maximizes in the
equivalent form sum_i |<i|r|i>|^2, since sum_i <i|r^2|i> is invariant under a
rotation among the occupied orbitals.
"""

import sys
import pathlib

import numpy as np

REPO = pathlib.Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "tools" / "cpu_validation"))
from gen_cpu_validation import bse_to_pyscf, molecule_form, CARTESIAN  # noqa: E402

ATOMS = [("O", (0.0, 0.0, 0.0)),
         ("H", (0.0, 0.0, 0.9584)),
         ("H", (0.9268, 0.0, -0.2400))]

#: How much PySCF may improve on our orbitals before we call them unconverged.
#: The Jacobi sweep stops on a 1e-10 angle, so the spread is stationary to far
#: better than this.
STATIONARY_TOL = 1e-7

def read_dump(path):
    tokens = path.read_text().split()
    basis = tokens[0]
    values = [float(x) for x in tokens[1:]]
    nao, n_occ = int(values[0]), int(values[1])
    at = 2
    l_before, l_after = values[at], values[at + 1]
    at += 2
    centroids = np.array(values[at:at + 3 * n_occ]).reshape((3, n_occ), order="F")
    at += 3 * n_occ
    orbitals = np.array(values[at:at + nao * n_occ]).reshape((nao, n_occ), order="F")
    return dict(basis=basis, nao=nao, n_occ=n_occ, l_before=l_before,
                l_after=l_after, centroids=centroids, orbitals=orbitals)


def main():
    from pyscf import gto, scf, lo

    failures = 0
    for index in (1, 2):
        path = pathlib.Path(f"/tmp/mqc_localize_{index}.txt")
        if not path.exists():
            print(f"  MISSING {path} -- run ./build/check_localize first")
            failures += 1
            continue
        ours = read_dump(path)

        symbols = {a[0] for a in ATOMS}
        mol = gto.Mole()
        mol.atom = ATOMS
        mol.unit = "Angstrom"
        mol.basis = {s: bse_to_pyscf(ours["basis"], s) for s in symbols}
        mol.cart = molecule_form(ours["basis"], symbols) == CARTESIAN
        mol.verbose = 0
        mol.build()

        mf = scf.RHF(mol)
        mf.conv_tol = 1e-11
        mf.kernel()
        n_occ = ours["n_occ"]

        def spread(coeff):
            """The PySCF Boys cost function: the sum of orbital spreads."""
            return float(lo.Boys(mol, coeff).cost_function(np.eye(coeff.shape[1])))

        ours_spread = spread(ours["orbitals"])

        # 1. Can PySCF improve on our orbitals? init_guess must be disabled --
        #    by default Boys discards whatever it was given in favour of an
        #    atomic guess, which is exactly why a naive comparison looks like a
        #    disagreement when there is none.
        refine = lo.Boys(mol, ours["orbitals"])
        refine.init_guess = None
        refine.conv_tol = 1e-12
        refine.max_cycle = 1000
        refined_spread = spread(refine.kernel())
        gain = ours_spread - refined_spread

        # 2. And what does PySCF reach by itself, for context?
        theirs = lo.Boys(mol, mf.mo_coeff[:, :n_occ])
        theirs.conv_tol = 1e-12
        theirs.max_cycle = 1000
        their_spread = spread(theirs.kernel())

        ok = gain < STATIONARY_TOL
        status = "ok  " if ok else "FAIL"
        print(f"  {status} {ours['basis']:8s} n_occ {n_occ}  spreads: "
              f"ours {ours_spread:.8f}   PySCF refining ours {refined_spread:.8f}"
              f"   PySCF alone {their_spread:.8f}")
        if not ok:
            failures += 1
            print(f"       PySCF improved our spread by {gain:.2e}, so ours is not "
                  f"a stationary point")
        elif their_spread < ours_spread - 1e-8:
            print(f"       note: PySCF localized better by "
                  f"{ours_spread - their_spread:.2e}")

        # Water should be recognisable: an oxygen core, two O-H bonds, and two
        # equivalent lone pairs off the molecular plane. Printed for a human; the
        # assertion above is what a machine checks.
        for i in range(n_occ):
            c = ours["centroids"][:, i]
            print(f"        centroid {i + 1}: "
                  f"({c[0]:8.4f},{c[1]:8.4f},{c[2]:8.4f})  |r| = "
                  f"{np.linalg.norm(c):7.4f}")

    print()
    if failures:
        print(f"[localize] {failures} FAILURE(S)")
        return 1
    print("[localize] our orbitals are stationary under the PySCF Boys objective")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: validation/check_localize.f90
!! Manual check that Boys localization is right
!!
!!     cmake -B build -DMQC_ENABLE_LIBCINT=ON
!!     ./build/check_localize && python3 validation/check_localize.py
!!
!! Milestone 2 of the EFP plan. Three internal checks that need no reference at
!! all, then a dump for comparison against `pyscf.lo.Boys`:
!!
!!   * **the energy is invariant.** A rotation among the occupied orbitals cannot
!!     change the density, so `Tr(D H) + ...` is untouched. This is the check that
!!     catches a rotation that is not orthogonal, which is the way a Jacobi sweep
!!     usually breaks, and it is exact rather than approximate.
!!   * **the functional rises monotonically.** The derivation says each pair's
!!     gain is `sqrt(P^2+Q^2) - P >= 0`, so no sweep may lower L. A wrong sign in
!!     the angle still converges -- to a stationary point that is a minimum or a
!!     saddle -- and the energy check would not notice, because any orthogonal
!!     rotation preserves it.
!!   * **the orbitals stay orthonormal**, `C^T S C = 1`.
!!
!! Then water's centroids, which should be recognisably two bonds and two lone
!! pairs, and are what the polarizabilities will later be placed on.
program check_localize
   use pic_types, only: dp
   use mqc_libcint_integrals, only: libcint_molecule_t, build_libcint_molecule
   use mqc_libcint_rhf, only: rhf_result_t, run_libcint_rhf
   use mqc_libcint_localize, only: boys_localize
   use mqc_error, only: error_t
   implicit none

   real(dp), parameter :: ANG = 1.8897261254578281_dp
   integer :: failures

   failures = 0
   call one_case(1, "sto-3g")
   call one_case(2, "cc-pvdz")

   write (*, "(A)") ""
   if (failures == 0) then
      write (*, "(A)") "[localize] wrote every case; run validation/check_localize.py"
   else
      write (*, "(A,I0,A)") "[localize] ", failures, " FAILURE(S)"
      error stop 1
   end if

contains

   subroutine one_case(case_index, basis)
      integer, intent(in) :: case_index
      character(len=*), intent(in) :: basis

      type(libcint_molecule_t) :: mol
      type(rhf_result_t) :: scf
      type(error_t) :: err
      real(dp), allocatable :: loc(:, :), cen(:, :), s(:, :), work(:, :), gram(:, :)
      real(dp), allocatable :: density(:, :)
      real(dp) :: c(3, 3)
      real(dp) :: l_before, l_after, worst_orth, e_before, e_after
      integer :: unit, i, j, k, n_occ, sweeps
      character(len=48) :: name

      c = reshape([0.0_dp, 0.0_dp, 0.0_dp, &
                   0.0_dp, 0.0_dp, 0.9584_dp*ANG, &
                   0.9268_dp*ANG, 0.0_dp, -0.2400_dp*ANG], [3, 3])

      call build_libcint_molecule([8, 1, 1], ["O ", "H ", "H "], c, basis, mol, err)
      if (err%has_error()) then
         write (*, "(A,A)") "[localize] basis failed: ", err%get_message()
         failures = failures + 1
         return
      end if

      call run_libcint_rhf(mol, 10, 100, 1.0e-11_dp, 1.0e-9_dp, .false., scf, err, &
                           in_core=.true.)
      if (err%has_error() .or. .not. scf%converged) then
         write (*, "(A)") "[localize] SCF failed"
         failures = failures + 1
         call mol%destroy()
         return
      end if
      n_occ = scf%n_occupied
      e_before = scf%energy

      ! The functional before localizing, from the canonical orbitals.
      call boys_localize(mol, scf%orbitals, n_occ, loc, cen, err, max_sweeps=0, &
                         functional=l_before)
      if (.not. err%has_error()) then
         deallocate (loc, cen)
         call boys_localize(mol, scf%orbitals, n_occ, loc, cen, err, &
                            sweeps_taken=sweeps, functional=l_after)
      end if
      if (err%has_error()) then
         write (*, "(A,A)") "[localize] failed: ", err%get_message()
         failures = failures + 1
         call mol%destroy()
         return
      end if

      ! Orthonormality: C^T S C must be the identity.
      call mol%overlap(s)
      allocate (work(mol%nao, n_occ), gram(n_occ, n_occ))
      work = matmul(s, loc)
      gram = matmul(transpose(loc), work)
      worst_orth = 0.0_dp
      do j = 1, n_occ
         do i = 1, n_occ
            if (i == j) then
               worst_orth = max(worst_orth, abs(gram(i, j) - 1.0_dp))
            else
               worst_orth = max(worst_orth, abs(gram(i, j)))
            end if
         end do
      end do

      ! The density from the localized orbitals must equal the SCF one, which is
      ! the statement that the energy did not move -- and is sharper than
      ! recomputing an energy, since it compares the object the energy is built
      ! from rather than a scalar that could agree by cancellation.
      allocate (density(mol%nao, mol%nao))
      density = 2.0_dp*matmul(loc, transpose(loc))
      e_after = maxval(abs(density - scf%density))

      write (name, "(A,I0,A)") "/tmp/mqc_localize_", case_index, ".txt"
      open (newunit=unit, file=trim(name), status="replace", action="write")
      write (unit, "(A)") trim(basis)
      write (unit, "(I0,1X,I0)") mol%nao, n_occ
      write (unit, "(es25.16e3)") l_before
      write (unit, "(es25.16e3)") l_after
      write (unit, "(3es25.16e3)") cen
      write (unit, "(es25.16e3)") loc
      close (unit)

      write (*, "(A,A,A,I0,A,I0,A,F12.6,A,F12.6,A,ES9.2,A,ES9.2)") &
         "  ", trim(basis), "  n_occ ", n_occ, "  sweeps ", sweeps, &
         "   L ", l_before, " -> ", l_after, &
         "   orth ", worst_orth, "   density ", e_after

      if (l_after < l_before - 1.0e-12_dp) then
         write (*, "(A)") "    FAIL the Boys functional went down"
         failures = failures + 1
      end if
      if (worst_orth > 1.0e-10_dp) then
         write (*, "(A)") "    FAIL the localized orbitals are not orthonormal"
         failures = failures + 1
      end if
      if (e_after > 1.0e-10_dp) then
         write (*, "(A)") "    FAIL localization changed the density"
         failures = failures + 1
      end if

      call mol%destroy()
      deallocate (loc, cen, s, work, gram, density)
   end subroutine one_case

end program check_localize
